common/src/main/java/com/wynntils/features/user/WorldWaypointDistanceFeature.java

```diff
@@ -82,61 +96,172 @@ public void onRenderLevelPost(RenderLevelEvent.Post event) {
+        }
     }
 
-    private Vector3d worldToNdc(Vector3f delta, Matrix4f projection) {
+    private Vec3 worldToScreen(Vector3f delta, Matrix4f projection) {
         Vector4f clipCoords = new Vector4f(delta.x(), delta.y(), delta.z(), 1.0f);
         clipCoords.transform(projection);
 
-        return new Vector3d(
+        // stands for Normalized Device Coordinates
+        Vector3d ndc = new Vector3d(
                 clipCoords.x() / clipCoords.w(), clipCoords.y() / clipCoords.w(), clipCoords.z() / clipCoords.w());
+
+        Window window = McUtils.window();
+
+        return new Vec3(
+                (float) ((ndc.x + 1.0f) / 2.0f) * window.getGuiScaledWidth(),
+                (float) ((1.0f - ndc.y) / 2.0f) * window.getGuiScaledHeight(),
+                (float) ndc.z);
+    }
+
+    // draw a line from screen center to the target's screenspace coordinate
+    // and find the intersect point on one of the screen's bounding
+    private Vec2 getBoundingIntersectPoint(Vec3 position, Window window) {
+        if (isInBound(position, window)) return null;
+        Vec3 centerPoint = new Vec3(window.getGuiScaledWidth() / 2, window.getGuiScaledHeight() / 2, 0);
+
+        // minecraft's origin point is top left corner
+        // so positive Y is at the screen bottom
+        float minX = (float) -(centerPoint.x - horizontalBoundingDistance);
+        float maxX = (float) centerPoint.x - horizontalBoundingDistance;
+        float minY = (float) -(centerPoint.y - topBoundingDistance);
+        float maxY = (float) centerPoint.y - bottomBoundingDistance;
+
+        // drag the origin point to center since indicator's screenspace position / rotation is in relation to it
+        Vec3 centerRelativePosition = screenCoord.subtract(centerPoint);
+
+        // invert xy axis if target is behind camera
+        if (centerRelativePosition.z > 1) {
+            centerRelativePosition = centerRelativePosition.multiply(-1, -1, 1);
+        }
+
+        // since center point is now the origin point, atan2 is used to get the angle, and angle is used to get the
+        // slope
+        double angle = StrictMath.atan2(centerRelativePosition.y, centerRelativePosition.x);
+        double m = StrictMath.tan(angle);
+
+        // trying to solve (y2 - y1) = m (x2 - x1) + c here
+        // starting from origin point/screen center (x1, y1), end at one of the screen bounding
+        // (y2 - y1) is the equivalent of the y position
+        // (x2 - x1) is the equivalent of the x position
+        // c is the line's y-intercept, but line pass through origin, so this will be 0
+        // finalize to y = mx, or x = y/m
+
+        if (centerRelativePosition.x > 0) {
+            centerRelativePosition = new Vec3(maxX, maxX * m, 0);
+        } else {
+            centerRelativePosition = new Vec3(minX, minX * m, 0);
+        }
+
+        if (centerRelativePosition.y > maxY) {
+            centerRelativePosition = new Vec3(maxY / m, maxY, 0);
```

Comment:
Are we 100% sure we can't get a division by zero here? If the angle is 0 then m will be 0. But I must admit I got a bit lost on the maths here.

Reply:
> Are we 100% sure we can't get a division by zero here? If the angle is 0 then m will be 0. But I must admit I got a bit lost on the maths here.

centerRelativePosition.y > maxY is the case where the line intersect with bottom bounding, in which the only case where slope is zero is probably where resolution's height being 0, I don't think such edge case exist in any playable situation

Reply:
on second thought, situation might be possible if bottom bounding goes higher than screen center, so it think it's intersecting with right and bottom bounding simultaneously, I'll fix it if I can get a crash out of it